db/mongodb/mongodb.go

```diff
@@ -0,0 +1,221 @@
+package mongodb
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+	"go.vocdoni.io/dvote/db"
+	"go.vocdoni.io/dvote/log"
+)
+
+const (
+	MONGODB_TIMEOUT_CONNECT = 10 * time.Second
+	MONGODB_TIMEOUT_COMMIT  = 10 * time.Second
+	MONGODB_TIMEOUT_QUERY   = 4 * time.Second
+)
+
+type MongoDB struct {
+	db         *mongo.Database
+	collection string
+}
+
+type KeyVal struct {
+	Key   string `bson:"_id" json:"key"` // needs to be an ugly string because of regex
+	Value []byte `bson:"value" json:"value"`
+}
+
+func New(opts db.Options) (*MongoDB, error) {
+	url := os.Getenv("MONGODB_URL")
+
+	if url == "" {
+		return nil, fmt.Errorf("missing MONGODB_URL env var")
+	}
+	// we use the path as the collection name (and we hash it to avoid invalid chars)
+	collection := fmt.Sprintf("%x", sha256.Sum256([]byte(opts.Path)))[:12]
+	log.Debugw("connecting to mongo database", "url", url, "collection", collection)
+	clientOptions := options.Client().ApplyURI(url).SetMaxConnecting(20)
+	ctx, cancel := context.WithTimeout(context.Background(), MONGODB_TIMEOUT_CONNECT)
+	defer cancel()
+	client, err := mongo.Connect(ctx, clientOptions)
+	if err != nil {
+		return nil, err
+	}
+	database := client.Database(collection)
+
+	// Shutdown database connection when SIGTERM received
+	go func() {
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt, syscall.SIGTERM)
```

Comment:
this is a bit hacky - a library package like db/mongodb shouldn't listen for a process signal, it should have a Shutdown method. it should only be the main package that does OS process stuff like listening for signals.